Tests for VSCode extension

chrmarti opened the pull request:
This is a set of tests to serve as a canary for when changes in the CLI will affect the VSCode extension (https://marketplace.visualstudio.com/items?itemName=ms-vscode.azurecli). It includes a module for the extension to consume and the tests to test against. This module encapsulates a few lines of code that seem to rely heavily on the CLI's implementation and are more likely to require updates with changes to the underlying parts of the CLI.

Not sure where this code should go, I've put it in azure.cli.core.tooling to get started.

troydai commented:
@chrmarti thanks for creating the pull requests. There are dependencies errors detected during the CI build, can you take a look? Since it has been for 16 days, can you rebase your changes?

troydai commented:
@chrmarti can you take a look your changes? It has been more than 3 weeks since it was submitted. We do require all the changes to be signed off and pass CI before we can accept it. We appreciate your contribution, but we do want to control the number of the open pull requests. Please update your pull request. We may close this pull request in a few days if there aren't any activities. If you're really busy but the change is essential to your product, please let me know.

chrmarti commented:
@troydai Thanks for following up, I was busy and am getting back to this only now.

I've rebased the changes and now two of the test runs fail. I've had a look at the failures and these seem to be in other tests than the ones I added. My changes shouldn't affect these and the error message looks like it is caused by something else: "Failure: CloudNotRegisteredException (The cloud 'AzureCloud' is not registered.)"

There have been other test runs, without my changes, that have failed with the same error message: https://travis-ci.org/Azure/azure-cli/jobs/239761829. Could this be an issue with the existing tests? /cc @chrisdias

troydai commented:
Thanks. The failures you notice coming from a parallelism conflict introduced by other changes recently. I'm rerunning the tests and hopefully, it will resolve the issues.

troydai commented on src/azure-cli-core/azure/cli/core/tooling.py:
```diff
@@ -0,0 +1,194 @@
+"""tooling integration"""
```
We don't require file level doc string. If you'd like to add, please move it behind the license header and add more details.

troydai replied:
And are these code production code or just serves the test purpose? If these are production code, I wonder if you can write your own wrap to hosting these code.

chrmarti replied:
`tooling.py` is production code. The reason I put it in the CLI is that this code relies on a lot of the CLI's details and having that on the CLI's side will allow for updating it together with the CLI, if it ever needs updating. Our tool, the VSCode extension, uses mostly `tooling.py` to access the CLI.

troydai commented:
Please forgive my reluctance of accepting this pull request at this point. I do think it is a very useful scenario. But I think there is an opportunity for us to draw a clear picture of the tooling extension story.

The tooling.py appears to me not to be an appropriate piece in the core, which we'd like to keep compact. Most of the code in the `tooling` modules seem to establish an alternative route to execute the azure-cli, which may lead to confusing scenarios.

The `tooling` module itself doesn't execute any code, which indicate that it is consumed by an outer python module. In that case, can't the out size python module host the `tooling` module?

If the underlying issue is the consistence and avoid breaking change, the test cases are good mechanism. If the requirement goes beyond that, I think a better approach is for us to meet and go through the whole picture.

/cc @derekbekoe

chrmarti commented:
I'm certainly fine with placing the module elsewhere in the CLI. I just wasn't sure where a good home would be.

This module is indeed consumed by VSCode extension. The point of hosting it in the CLI is that we can update it should any of its assumptions break and release that together with the breaking changes as part of the CLI. With such an approach the VSCode extension not (or less often) have to update in anticipation of breaking changes in the CLI. The VSCode extension will also not (or less often) have to add version detection to handle different versions of the CLI.

troydai commented:
Per our discussion please:
1. Rebase on the latest master
2. Add comments stating the use case of this module and its status of deprecating.

troydai commented:
@chrmarti please let me know when I should begin reviewing.

chrmarti commented:
@troydai Looks like all checks pass now, please take a look. The only change to before is the doc string at the start of the file, hope that's what you had in mind.

troydai commented:
@tjprescott @derekbekoe how much of the functions will be moved to https://github.com/microsoft/knack here? I'm ready to accept this pull request.

chrmarti commented:
@troydai Anything I can do to help move this forward?

derekbekoe commented on src/azure-cli-core/azure/cli/core/tooling.py:
```diff
@@ -0,0 +1,199 @@
+    return command_table
+
+
+def _install_modules(command_table):
```
Once the extensions PR gets merged, this may have to change to handle loading of extensions as well. https://github.com/Azure/azure-cli/pull/4167/files#diff-34546b027e108d42bd46a0c08a1fe155R379

tjprescott commented on src/azure-cli-core/azure/cli/core/tooling.py:
```diff
@@ -0,0 +1,199 @@
+Tooling Integration
+
+This module is used by the Azure CLI Tools extension for VSCode. It should not be used
+anywhere else and will be removed in a future release.
```
Why will this be removed in a future version?

chrmarti replied:
My understanding is that there will be something (like a shell-core module) replacing this in the future and we agreed on this interim solution. @troydai Has the details on the CLI side.

tjprescott commented on src/azure-cli-core/azure/cli/core/tooling.py:
```diff
@@ -0,0 +1,199 @@
+        'options': ['--query'],
+        'help': 'JMESPath query string. See http://jmespath.org/ for more information and examples.'
+    }
+}
```
Is there not a programmatic way to retrieve these? Otherwise if a global argument were added, it risks not being added here.

chrmarti replied:
Absolutely, if this is not good enough for this interim solution, I'm happy to update it, but would appreciate some help / code pointers to look into.

troydai commented:
@chrmarti, I'm going to close this PR for now since there are different plans for our core code since our last discussion. I believe @tjprescott had communicated with you of our changes and which will eventually break your plugin once it is made. Let's work together to ensure the plugin will keep working with the new CLI core design off-line.

Thanks.
